backends/p4tools/modules/testgen/core/small_step/extern_stepper.cpp

```diff
@@ -210,21 +196,14 @@ void ExprStepper::evalInternalExternMethodCall(const IR::MethodCallExpression *c
              auto &nextState = state.clone();
              const auto *appendType = state.resolveType(appendVar->type);
 
-             if (appendType->is<IR::Type_StructLike>()) {
-                 auto appendRef = ToolsVariables::convertReference(appendVar);
-                 // We only support flat assignments, so retrieve all fields from the input
-                 // argument.
-                 const auto flatFields = nextState.getFlatFields(appendRef);
-                 for (const auto &fieldRef : flatFields) {
-                     nextState.appendToPacketBuffer(nextState.get(fieldRef));
+             nextState.add(*new TraceEvents::Expression(appendVar, "AppendToProgramHeader"));
+             // append the field to the packet buffer.
+             if (const auto *structExpr = appendVar->to<IR::StructExpression>()) {
+                 auto exprList = IR::flattenStructExpression(structExpr);
+                 for (const auto *expr : exprList) {
+                     nextState.appendToPacketBuffer(expr);
                  }
              } else if (appendType->is<IR::Type_Bits>()) {
-                 if (!(appendVar->is<IR::Member>() || appendVar->is<IR::PathExpression>() ||
-                       appendVar->is<IR::TaintExpression>() || appendVar->is<IR::Constant>())) {
-                     TESTGEN_UNIMPLEMENTED("append input %1% of type %2% not supported", appendVar,
-                                           appendVar->type);
-                 }
-                 nextState.add(*new TraceEvents::Expression(appendVar, "AppendToProgramHeader"));
```

Comment:
This piece of code is very similar to the one in prepend, could you please factor it out?

Reply:
I was kind of following the rule of three here. It did not seem worthwhile to factor it out because I would need to parametrize the append/prepend including the trace event. You end up saving not much and just gain more indirection.